clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleRegionalServerGroupCachingAgent.groovy

```diff
@@ -0,0 +1,545 @@
+
+      def instanceKeys = []
+      def loadBalancerKeys = []
+
+      cacheResultBuilder.namespace(APPLICATIONS.ns).keep(appKey).with {
+        attributes.name = applicationName
+        relationships[CLUSTERS.ns].add(clusterKey)
+      }
+
+      cacheResultBuilder.namespace(CLUSTERS.ns).keep(clusterKey).with {
+        attributes.name = clusterName
+        attributes.accountName = accountName
+        relationships[APPLICATIONS.ns].add(appKey)
+        relationships[SERVER_GROUPS.ns].add(serverGroupKey)
+      }
+
+      serverGroup.instances.each { GoogleInstance partialInstance ->
+        def instanceKey = Keys.getInstanceKey(accountName, serverGroup.region, partialInstance.name)
+        instanceKeys << instanceKey
+        cacheResultBuilder.namespace(INSTANCES.ns).keep(instanceKey).relationships[SERVER_GROUPS.ns].add(serverGroupKey)
+      }
+      serverGroup.instances.clear()
+
+      serverGroup.asg.loadBalancerNames.each { String loadBalancerName ->
+        loadBalancerKeys << Keys.getLoadBalancerKey(region,
+                                                    accountName,
+                                                    loadBalancerName)
+      }
+
+      loadBalancerKeys.each { String loadBalancerKey ->
+        cacheResultBuilder.namespace(LOAD_BALANCERS.ns).keep(loadBalancerKey).with {
+          relationships[SERVER_GROUPS.ns].add(serverGroupKey)
+        }
+      }
+
+      if (shouldUseOnDemandData(cacheResultBuilder, serverGroup)) {
+        moveOnDemandDataToNamespace(cacheResultBuilder, serverGroup)
+      } else {
+        cacheResultBuilder.namespace(SERVER_GROUPS.ns).keep(serverGroupKey).with {
+          attributes = objectMapper.convertValue(serverGroup, ATTRIBUTES)
+          relationships[APPLICATIONS.ns].add(appKey)
+          relationships[CLUSTERS.ns].add(clusterKey)
+          relationships[INSTANCES.ns].addAll(instanceKeys)
+          relationships[LOAD_BALANCERS.ns].addAll(loadBalancerKeys)
+        }
+      }
+    }
+
+    log.info("Caching ${cacheResultBuilder.namespace(APPLICATIONS.ns).keepSize()} applications in ${agentType}")
+    log.info("Caching ${cacheResultBuilder.namespace(CLUSTERS.ns).keepSize()} clusters in ${agentType}")
+    log.info("Caching ${cacheResultBuilder.namespace(SERVER_GROUPS.ns).keepSize()} server groups in ${agentType}")
+    log.info("Caching ${cacheResultBuilder.namespace(INSTANCES.ns).keepSize()} instance relationships in ${agentType}")
+    log.info("Caching ${cacheResultBuilder.namespace(LOAD_BALANCERS.ns).keepSize()} load balancer relationships in ${agentType}")
+    log.info("Caching ${cacheResultBuilder.onDemand.toKeep.size()} onDemand entries in ${agentType}")
+    log.info("Evicting ${cacheResultBuilder.onDemand.toEvict.size()} onDemand entries in ${agentType}")
+
+    cacheResultBuilder.build()
+  }
+
+  boolean shouldUseOnDemandData(CacheResultBuilder cacheResultBuilder, GoogleServerGroup googleServerGroup) {
```

Comment:
Probably identical to existing code, maybe pull this out into a cache utility class?